File: classified/packages/server/src/db/connection.ts
import { drizzle } from 'drizzle-orm/node-postgres';
import { drizzle as drizzlePglite } from 'drizzle-orm/pglite';
import { Pool } from 'pg';
import { PGlite } from '@electric-sql/pglite';
import { logger } from '@elizaos/core';
import * as schema from './schema';

export type DatabaseConnection = ReturnType<typeof drizzle> | ReturnType<typeof drizzlePglite>;

interface ConnectionOptions {
  postgresUrl?: string;
  dataDir?: string;
}

let db: DatabaseConnection | null = null;
let pgPool: Pool | null = null;
let pgliteConnection: PGlite | null = null;

export async function createDatabaseConnection(
  options: ConnectionOptions
): Promise<DatabaseConnection> {
  if (db) {
    return db;
  }

  if (options.postgresUrl) {
    // PostgreSQL connection
    logger.info('[MessageDB] Connecting to PostgreSQL database...');
    pgPool = new Pool({
      connectionString: options.postgresUrl,
      max: 10,
      idleTimeoutMillis: 30000,
      connectionTimeoutMillis: 30000,
    });

    db = drizzle(pgPool, { schema });
    logger.success('[MessageDB] Connected to PostgreSQL database');
  } else {
    // PGLite connection (default)
    const dataDir = options.dataDir || '.eliza/message-server/db';
    logger.info(`[MessageDB] Initializing PGLite database at: ${dataDir}`);

    pgliteConnection = new PGlite(dataDir);
    await pgliteConnection.waitReady;

    db = drizzlePglite(pgliteConnection, { schema });
    logger.success('[MessageDB] PGLite database initialized');
  }

  return db;
}

export function getDatabase(): DatabaseConnection {
  if (!db) {
    throw new Error('Database not initialized. Call createDatabaseConnection first.');
  }
  return db;
}

export async function closeDatabase(): Promise<void> {
  if (pgPool) {
    await pgPool.end();
    pgPool = null;
  }

  if (pgliteConnection) {
    await pgliteConnection.close();
    pgliteConnection = null;
  }

  db = null;
  logger.info('[MessageDB] Database connection closed');
}

File: cpp_merge/classified/packages/server/src/db/connection.cpp
#include "/home/runner/work/elizaos-cpp/elizaos-cpp/classified/packages/server/src/db/connection.h"

std::shared_ptr<Promise<DatabaseConnection>> createDatabaseConnection(std::shared_ptr<ConnectionOptions> options)
{
    if (db) {
        return db;
    }
    if (options->postgresUrl) {
        logger->info(std:("[MessageDB] Connecting to PostgreSQL database..."));
        pgPool = std::make_shared<Pool>(object{
            object::pair{std:("connectionString"), options->postgresUrl}, 
            object::pair{std:("max"), 10}, 
            object::pair{std:("idleTimeoutMillis"), 30000}, 
            object::pair{std:("connectionTimeoutMillis"), 30000}
        });
        db = drizzle(pgPool, object{
            object::pair{std:("schema"), std:("schema")}
        });
        logger->success(std:("[MessageDB] Connected to PostgreSQL database"));
    } else {
        auto dataDir = OR((options->dataDir), (std:(".eliza/message-server/db")));
        logger->info(std:("[MessageDB] Initializing PGLite database at: ") + dataDir + string_empty);
        pgliteConnection = std::make_shared<PGlite>(dataDir);
        std::async([=]() { pgliteConnection["waitReady"]; });
        db = drizzlePglite(pgliteConnection, object{
            object::pair{std:("schema"), std:("schema")}
        });
        logger->success(std:("[MessageDB] PGLite database initialized"));
    }
    return db;
};


DatabaseConnection getDatabase()
{
    if (!db) {
        throw any(std::make_shared<Error>(std:("Database not initialized. Call createDatabaseConnection first.")));
    }
    return db;
};


std::shared_ptr<Promise<void>> closeDatabase()
{
    if (pgPool) {
        std::async([=]() { pgPool["end"](); });
        pgPool = nullptr;
    }
    if (pgliteConnection) {
        std::async([=]() { pgliteConnection["close"](); });
        pgliteConnection = nullptr;
    }
    db = nullptr;
    logger->info(std:("[MessageDB] Database connection closed"));
    return std::shared_ptr<Promise<void>>();
};


any db = nullptr;
any pgPool = nullptr;
any pgliteConnection = nullptr;

void Main(void)
{
}

MAIN
